src/main/scala/org/kframework/minikore/MiniKore.scala

```diff
@@ -1,39 +1,310 @@
+    def unapply(arg: SortDeclaration): Option[(String, Attributes)] = Some((arg.sort, arg.att))
+  }
+
+  trait SymbolDeclaration extends Sentence {
+    val sort: String
+    val label: String
+    val args: Seq[String]
+    val att: Attributes
+  }
+
+  object SymbolDeclaration {
+    def unapply(arg: SymbolDeclaration): Option[(String, String, Seq[String], Attributes)] = Some((arg.sort, arg.label, arg.args, arg.att))
+  }
+
+  trait Rule extends Sentence {
+    val p: Pattern
+    val att: Attributes
+  }
+
+  object Rule {
+    def unapply(arg: Rule): Option[(Pattern, Attributes)] = Some((arg.p, arg.att))
+  }
+
 
-  sealed trait Sentence
-  case class Import(name: String, att: Attributes) extends Sentence
-  case class SortDeclaration(sort: String, att: Attributes) extends Sentence
-  case class SymbolDeclaration(sort: String, label: String, args: Seq[String], att: Attributes) extends Sentence
-  case class Rule(pattern: Pattern, att: Attributes) extends Sentence
-  case class Axiom(pattern: Pattern, att: Attributes) extends Sentence
+  trait Axiom extends Sentence {
+    val p: Pattern
+    val att: Attributes
+  }
+
+  object Axiom {
+    def unapply(arg: Axiom): Option[(Pattern, Attributes)] = Some((arg.p, arg.att))
+  }
+
+
+  // Pattern trait + traits
+  trait Pattern extends AST
+
+  trait Variable extends Pattern with Leaf[(String, String)]{
+    val name: String
+    val sort: String
+  }
+
+  object Variable {
+    def unapply(arg: Variable): Option[(String, String)] = Some((arg.name, arg.sort))
+  }
+
+  trait Application extends Pattern with Node[String] {
+    val label: String
+    val args: Seq[Pattern]
+  }
+
+  object Application extends Pattern {
+    def unapply(arg: Application): Option[(String, Seq[Pattern])] = Some((arg.label, arg.args))
+  }
+
+  trait DomainValue extends Pattern with Leaf[(String, String)] {
+    val label: String
+    val value: String
+  }
+
+  object DomainValue extends Pattern {
+    def unapply(arg: DomainValue): Option[(String, String)] = Some((arg.label, arg.value))
+  }
 
-  sealed trait Pattern
-  case class Variable(name: String, sort: String) extends Pattern
-  case class Application(label: String, args: Seq[Pattern]) extends Pattern
-  case class DomainValue(label: String, value: String) extends Pattern
   //
-  case class True() extends Pattern
-  case class False() extends Pattern
+  trait True extends Pattern with Leaf[Boolean]
+
+  object True {
+    def unapply(arg: True): Boolean = true
+  }
+
+  trait False extends Pattern with Leaf[Boolean]
+
+  object False {
+    def unapply(arg: False): Boolean = true
+  }
+
+
   //
-  case class And(p: Pattern, q: Pattern) extends Pattern
-  case class Or(p: Pattern, q: Pattern) extends Pattern
-  case class Not(p: Pattern) extends Pattern
-  case class Implies(p: Pattern, q: Pattern) extends Pattern
-  case class Exists(v: Variable, p: Pattern) extends Pattern
-  case class ForAll(v: Variable, p: Pattern) extends Pattern
```

Comment:
And extends both pattern and Node. So suppose patterns x, y, and p = Pattern(x, y). Now, you can write stuff like - 
p match {case Pattern(A:Pattern, B: Pattern) => ..} or

p match {case Node("\and", Seq(A: AST, B: AST)) => ...}.